File: login-server/src/main/java/work/fking/pangya/login/Player.java
package work.fking.pangya.login;

import io.netty.channel.Channel;
import work.fking.pangya.common.Rand;
import work.fking.pangya.login.auth.UserInfo;
import work.fking.pangya.login.net.LoginState;

public class Player {

    private final int uid;
    private final Channel channel;

    private final String loginKey;
    private final String sessionKey;
    private final String username;
    private final String nickname;

    private LoginState loginState = LoginState.AUTHENTICATED;

    public Player(Channel channel, UserInfo userInfo) {
        this.channel = channel;
        loginKey = Rand.randomHexString(16);
        sessionKey = Rand.randomHexString(16);
        uid = userInfo.uid();
        username = userInfo.username();
        nickname = userInfo.nickname();
    }

    public int uid() {
        return uid;
    }

    public Channel channel() {
        return channel;
    }

    public String loginKey() {
        return loginKey;
    }

    public String sessionKey() {
        return sessionKey;
    }

    public String username() {
        return username;
    }

    public String nickname() {
        return nickname;
    }

    public LoginState loginState() {
        return loginState;
    }

    public void setLoginState(LoginState loginState) {
        if (this.loginState.validTransition(loginState)) {
            this.loginState = loginState;
        } else {
            throw new IllegalStateException("Attempted to set loginState to " + loginState + " from " + this.loginState);
        }
    }
}

File: login-server/src/main/kotlin/work/fking/pangya/login/Player.kt
package work.fking.pangya.login

import io.netty.channel.Channel
import work.fking.pangya.login.auth.UserInfo
import work.fking.pangya.login.net.LoginState
import work.fking.pangya.login.net.LoginState.AUTHENTICATED

class Player(
    private val channel: Channel,
    val uid: Int,
    val loginKey: String = Rand.randomHexString(16),
    val sessionKey: String = Rand.randomHexString(16),
    val username: String,
    var nickname: String?,
    val hasBaseCharacter: Boolean = false
) {
    var state: LoginState = AUTHENTICATED
        set(value) {
            require(state.validTransition(value)) { "Invalid loginState transition, from $state to $value" }
            field = value
        }

    // these values are passed through the SessionInfo for the Game Server to create the desired starting character.
    // It is done like this to avoid having to copy over character creation logic from the game server code base.
    var pickedCharacterIffId: Int? = null
    var pickedCharacterHairColor: Int? = null

    constructor(channel: Channel, userInfo: UserInfo) : this(
        channel = channel,
        uid = userInfo.uid,
        username = userInfo.username,
        nickname = userInfo.nickname,
        hasBaseCharacter = userInfo.hasBaseCharacter
    )

    fun write(message: Any) {
        channel.write(message, channel.voidPromise())
    }

    fun writeAndFlush(message: Any) {
        channel.writeAndFlush(message)
    }

    fun flush() {
        channel.flush()
    }
}
